p2p/net/swarm/clock.go

```diff
@@ -0,0 +1,49 @@
+package swarm
+
+import "time"
+
+// InstantTimer is a timer that triggers at some instant rather than some duration
+type InstantTimer interface {
+	Reset(d time.Time) bool
+	Stop() bool
+	Ch() <-chan time.Time
+}
+
+// Clock is a clock that can create timers that trigger at some
+// instant rather than some duration
+type Clock interface {
```

Comment:
we don't need a new interface. 
In this specific case I'm using `InstantTimer` which is not provided by benbjohnson/clock, but I can use standard Timers. I didn't use benbjohnson/clock  because of the negative timer not being fired immediately and I thought we were going to use our own implementation going forward.  

I see that https://github.com/benbjohnson/clock/pull/50 is merged. So I don't have any objections to using benbjohnson/clock.  

@MarcoPolo what do you think?

Reply:
The fix was released in https://github.com/benbjohnson/clock/releases/tag/v1.3.4.

Reply:
If you are setting a timer based on an instant in time rather than some duration you should use this clock. The benbjohnson clock will be flaky for this use case because you have two Goroutines that are both trying to use the value returned from `Now()`. 

Here's a simple example: that library has an internal ticker and you have your timer handler logic. Your handler wants to reset the timer for 1 minute from the moment it was called (now), and after the library has finished notifying all timers, it'll advance the clock (let's call the advanced clock time the future). If the your handler goroutine calls reset before the ticker finishes calling all timers and advancing the clock, you're fine because now the timer is registered for now+1min. But if the ticker advanced to the future you're out of luck because you've just registered the timer for the future+1min.

This isn't a problem with the benbjohnson clock, it's actually a problem with trying to mock the timer interface since this only accepts a duration not a time. Which is why this `Clock` interface lets you define timers that trigger at some point in time rather than by some duration in the future.

Does that make sense? If so I think we should include this logic in the codebase as a comment when this comes up again in the future, since it's not super obvious.


Another added bonus is that this mock clock can be implemented in about 100 LoC :)

Reply:
Thanks @MarcoPolo. I didn't realise this case would be flaky. We should add this comment.